Use TimestampData to keep track of the progress of coverage providers, monitors, and scripts

leonardr opened the pull request:
This branch introduces a new object to the metadata layer: `TimestampData`. This object corresponds to a row in the `timestamps` table.

`Script`, `Monitor`, and `CoverageProvider` all have an architecture where the superclass calls a "do something" method that the subclass is supposed to define. The `TimestampData` object is now how the subclass communicates back to the superclass about what happened in the "do something" method. The superclass is responsible for filling in any missing information from the `TimestampData` and actually writing it to the database.

In the simplest case, where the "do something" method doesn't have any particular status to report, `Script`, `Monitor`, and `CoverageProvider` will capture the start and finish times, record any exceptions, and write the `TimestampData` to the database.

In more complicated cases, any subclass of `Script`, `Monitor`, or `CoverageProvider` has a significant ability to  customize what goes into its `timestamps` row by having its 'do something' method return a `TimestampData`.

`Monitor` is a little different because if there's an unhandled exception we don't want to modify the `start` or `finish` times. So that class still makes direct use of the `Timestamp` object a lot.

`CoverageProvider` is the complicated one. I subclassed `TimestampData` as `CoverageProviderProgress` so that I could use it to keep track of the variable `offset`. This is different from `TimestampData.counter`, which is written to the database and kept across runs. The `offset` is used _within_ a run to skip over certain database rows that are known to be ignorable _for this run_.

(This may seem like overkill, but my next branch is going to add to `CoverageProviderProgress` the ability to keep track of successes and failures _across batches_, so it's really useful to have a single object that's being passed back and forth between calls to `run_once`.)

leonardr commented on coverage.py:
```diff
@@ -174,7 +189,12 @@ def operation(self):
+
+        result = result or CoverageProviderProgress()
+        self.finalize_timestampdata(result, start=start)
+        return result
```
This pattern will also show up in Monitor and Script. The generic method calls the "do something" method, expects a TimestampData back, and writes the TimestampData to the database.

leonardr commented on metadata_layer.py:
```diff
@@ -502,6 +504,128 @@ def __init__(self, content_type, drm_scheme, link=None, rights_uri=None):
             self.rights_uri = self.link.rights_uri
 
 
+class TimestampData(object):
```
This is the object itself. Some pieces of information, like `service_type` and `collection_id`, are provided by the superclass service code -- some CoverageProvider subclass can't decide that it's actually a script. 

Other pieces of information, like `start` and `finish` and `counter`, can be provided by the subclassed "do something" method, but the superclass method will provide default values if they're not provided.

Other pieces of information, like `achievements`, are ignored unless the "do something" method sets them.

`exception` is set by the superclass if the "do something" method raises an exception, but theoretically the "do something method" can set it directly and everything should work.

leonardr commented on monitor.py:
```diff
@@ -293,49 +309,45 @@ def __init__(self, _db, collection=None, batch_size=None):
-            start=run_started_at, finish=run_ended_at,
-            achievements=achievements, counter=new_offset,
-            exception=exception
-        )
```
We no longer need to update the timestamp or catch an exception in this method -- we return a TimestampData with the new `counter` if we succeed, and if we raise an exception, `run` will catch it.

leonardr commented on monitor.py:
```diff
@@ -293,49 +309,45 @@ def __init__(self, _db, collection=None, batch_size=None):
         super(SweepMonitor, self).__init__(_db, collection=collection)
 
-    def run(self):
+    def run_once(self, *ignore):
```
Bu subclassing `run_once` (the "do something" method) instead of `run`, we get to reuse the code that manages the Timestamp and handles exceptions. That's why code to do those two things is removed from this method.

EdwinGuzman commented on metadata_layer.py:
```diff
@@ -502,6 +504,128 @@ def __init__(self, content_type, drm_scheme, link=None, rights_uri=None):
+            if finish is self.NO_VALUE:
+                finish = datetime.datetime.utcnow()
+            self.finish = finish
+        if self.start is self.NO_VALUE:
```
This check is being done above. This is the case where both instances of `start` are `NO_VALUE`? And why do you need to set it to `self.finish`?

leonardr replied:
Yes, exactly -- neither the constructor nor `finalize` explained  when the service started running. In that case we use the only date we have available -- the time when the service started running. If it's also the case that no one explained when the service _stopped_ running (which is pretty likely) then we use the current time for both dates.

The goal is just to have both dates set to some value for all `Timestamp` objects, so that we can do things like sort by one field or another in database queries.

leonardr commented on monitor.py:
```diff
@@ -293,49 +309,45 @@ def __init__(self, _db, collection=None, batch_size=None):
-        achievements = None
+        timestamp.start = run_started_at
+
         while True:
```
It'll finish if there's an exception (we catch the exception and `break`) or if `process_batch` returns 0 (indicating that the sweep has completed and the database ID reset to zero).

leonardr commented:
Sorry for the mess; this ended up more complicated than I hoped.

* Created Timestamp.to_data, which converts a database Timestamp object to an unfinalized TimestampData.
* Defined DatabaseTest.time_eq, which makes it really easy to assert that two datetimes are _almost_ equivalent -- they don't have to be exactly the same.
* Created TimelineMonitor, which abstracts away a very common pattern in monitors.
* Moved some code out of Monitor.timestamp to Monitor.initial_start_time so that it could be used in TimelineMonitor.

Finally, this is the big one: I changed `Monitor.run_once` to take a TimestampData object instead of a `start` and `cutoff` time. If run_once returns a TimestampData, that's the one we finalize and write to the `timestamps` table. This gives a Monitor subclass a lot of flexibility to record its status in TimestampData.achievements, mess around with what the dates in the timestamps mean, etc.

TimelineMonitor brings back the start/cutoff behavior, which let me minimize the changes necessary in circulation. Instead of making a big change to the `run_once` implementations, I changed monitors to subclass `TimelineMonitor`, and renamed `run_once` to `catch_up_from`.

EdwinGuzman commented on model/coverage.py:
```diff
@@ -242,6 +242,13 @@ def update(self, start=None, finish=None, achievements=None,
 
+    def to_data(self):
+        """Convert this Timestamp to an unfinalized TimestampData."""
+        from ..metadata_layer import TimestampData
```
What does the double dots mean for this import statement?

leonardr replied:
We're in the `core.model` package, and we need to import something from the `core` package above. It's analogous to `..` in a shell referring to the parent directory.

This is also why the import statement is inside the function (executed at runtime), rather than at the top of the file (executed on startup). `metadata_layer.py` requires model code to run, and trying to import `metadata_layer` while the model code is loading would create a loop.

EdwinGuzman commented on model/coverage.py:
```diff
@@ -242,6 +242,13 @@ def update(self, start=None, finish=None, achievements=None,
         self.exception = exception
 
+    def to_data(self):
+        """Convert this Timestamp to an unfinalized TimestampData."""
```
Unfinalized means that it hasn't finished running so what values should it not have until it is finalized?

leonardr replied:
'Unfinalized' means it's missing the information about which service it's a timestamp for. In this case we're deliberately leaving that data out, and that's because I want to force `finalize` to be called on this timestamp before it's written to the database. And _that's_ because `finalize` does other stuff like fill in defaults for any missing values.

EdwinGuzman commented on model/coverage.py:
```diff
@@ -100,6 +100,11 @@ class Timestamp(Base):
+    # A stand-in value used to indicate that a field in the timestamps
+    # should be set to None. This is necessary because 'None' generally
+    # means 'use the default value'.
+    NO_VALUE = object()
```
Why/when would you want a Timestamp.finish to be able to take on a null value?

leonardr replied:
In most cases it's not necessary. The motivating case is `MWCollectionUpdateMonitor` in circulation, which uses `timestamps.finish` to ask for changes since a certain time. If that monitor gets an empty OPDS feed, we want to leave `timestamps.finish` alone, not update it to the current time.

This is to avoid the possibility that there are pending changes that didn't show up the last time we asked for changes since time X, but will show up the next time we ask.

In the case where `timestamps.finish` starts out None, we want to make sure it stays None.

The BibliothecaEventMonitor used to be the same way -- it used `timestamps.start` and `timestamps.finish` to track the timestamps of the first and last events it saw in a given run. However, that was only for informational purposes, and it wasn't really that interesting, so I was able to convert that monitor over to a TimelineMonitor.

(It might not be that bad to also convert MWCollectionUpdateMonitor, but missing an update from the metadata wrangler is worse than missing one from Bibliotheca, because we have no metadata wrangler equivalent of the BibliothecaCirculationSweep.)
